### PR TITLE
reworks bootstrap/cmake options for building Epetra vs Tpetra

### LINKED ISSUE #585: change bootstrap and cmake --enable-kokkos option to --enable-tpetra
Currently Kokkos gets built whether --enable-kokkos is true or false.  This flag is intended to mean that the code is built with Tpetra and not Epetra.  

Change this flag in both bootstrap and CMake (on main branch) to --enable-tpetra.  Add new flag --enable-epetra.  Make sure that --enable-epetra and --enable-tpetra can coexist (potentially without CUDA and/or OpenMP).

Keep the --enable-kokkos_cuda and --enable-kokkos_openmp flags -- these logically mean something, but do check all combinations make are the right choices in our cmake.

### PR DESCRIPTION
**Pull request by ecoon**
bootstrap option, cmake option, what it does

* --enable-epetra; ENABLE_Epetra; turns on Epetra branch of code in TPLs
* --enable-tpetra; ENABLE_Tpetra; turns on Tpetra branch of code in TPLs
* --enable-cuda; ENABLE_CUDA; calls find_package(CUDA), adds extra Kokkos arch, enables in Hypre
* --enable-openmp; ENABLE_OpenMP; calls find_package(OpenMP), adds Kokkos arch, enables in Hypre

--enable-kokkos is now gone (as Kokkos was built even when --disable-kokkos was set)

This will close #585 

This is a WIP.  It must be tested on:

- [x] Epetra only + serial, builds master
- [x] Tpetra only + serial, builds tpetra-solvers
- [x] Epetra + Tpetra TPLs that can build both master (Epetra) and tpetra-solvers (Tpetra)
- [x] Tpetra + CUDA @JulienLoiseau 
- [x] Tpetra + OpenMP @JulienLoiseau 
- [x] Epetra + OpenMP -- what does or should this do?  Do we want to support this? @lipnikov 
- [x] Structured (@ecoon )
- ~~Structured + OpenMP -- can/should this work?  It looks like it did at one point but then was commented out?~~ untested in current master, so removing this requirement

@jd-moulton David, I reorganized Trilinos packages a bit.  In particular, the list of things to check for in AmanziTPLs.cmake was not the full set of things built in Build_Trilinos, and was missing some things that we needed.  So those code are identical now.  Except that Basker is a header-only library and doesn't provide the CMake target info, so we cannot find_package(Basker).  At any rate, I reorganized some of the TPL logic, so please take a look and see if you find anything I messed up.

@lipnikov I have been able to use 1 TPLs build with both master and tpetra-solvers now.  This was done only with serial though -- I'm not sure what would happen if you tried to use OpenMP.  I'm not sure whether we've tested Epetra + Hypre OpenMP in a very long time, but conceptually it should work.

@JulienLoiseau This likely will need to merge with your PR for Hypre.  Let me know if there are things you'd like me to change here to make that easier.  I got at least your ENABLE_HYPRE change included here, but there may be some merges in Build_HYPRE or similar.

Note I also found a few extra flags in ShyLU and Amesos2.  Most interestingly, there is an Amesos2_ENABLE_CUSPARSE option that I turned on for us to try.  Also, there is a FastILU in ShyLU that I think I have turned on and should get built now in ShyLU_Node.  Not sure if @hlim88 is looking at this yet or not, but it should be accessible through Ifpack2 as the Filu preconditioner.

### CODE REVIEW

**Comment by hlim88** [comment]
@ecoon Yes, I am looking `FIlu` in `Shylu` which can be directly called by `ifpack2` under `Details`. I tested some basics and I cannot find different performances between `FIlu` and `ILUT`. 

Also we need to make sure that having `Shylu_node` not just `Shylu_dd` (which @JulienLoiseau mentioned to me).

**Comment by ecoon** [comment]
@hlim88 That's a little disappointing, considering the ILUT was host-only code if I remember right.  Assuming they aren't actually the same thing or even were actually the same code, one would expect any other implementation that uses Kokkos to be better.

**Comment by lipnikov** [comment]
Epetra + OpenMP, master branch.

verify_operators "Hypre: AMG" structured3d 40 fv 1e-10

FV took 5 iterations. MFD (4, 6, 58 iterations) ???

np = 1, solution time = 2.53 (FV)   1.80 (MFD, per itrs)
np = 2, solution time = 1.84 (FV)   1.10 (MFD, per itrs)
np =4, solution time = 1.25 (FV)    0.69 (MFD, per itrs)

**Comment by JulienLoiseau** [comment]
Tpetra + OpenMP, tpetra branch. 

srun -n X ./src/operators/verify_operators "Hypre: AMG" structured3d 70 fv 1e-10 1 gmres omp

Solver time. 

MPI - OMP (1): 
1: 32.3947
2: 25.6081
4: 17.9367

MPI (1) - OMP: (need to work on srun options)
1: 32.5867
2: 26.7223
4: 23.0053

File: tools/cmake/AmanziTPL.cmake [hunk 1]
```diff
@@ -655,19 +616,53 @@ if (ENABLE_ALQUIMIA)
   endif()
 endif()
 
+
+##############################################################################
+# PETSc - http://www.mcs.anl.gov/petsc
+##############################################################################
+option(ENABLE_PETSC "Enable PETSc for use in structured or Alquimia" FALSE)
+# note, this is a bit off -- Alquimia doesn't require PETSc, PFLOTRAN
```
Comment (lipnikov) [inline tools/cmake/AmanziTPL.cmake:624]: To implement interface to PFLotran, Alquimia uses PETSc headers.

**Review by lipnikov — COMMENTED** [review]
I see no obvious issues, but additional testing for various combinations of input options is required.

**Comment by JulienLoiseau** [comment]
I had to make yet another patch for ifpack2. Indeed before we were using our "own" version which also fixed a but of multiple Backends for Kokkos (OMP+Serial, CUDA+Serial or CUDA+OMP+Serial). 
This also required a fix in the tpetra branch itself which now builds and test properly for both CUDA and OMP. 
The Serial backend is now always ON in Build_Trilinos.cmake.

**Comment by JulienLoiseau** [comment]
Tried to build on Summit with Tpetra + CUDA + Arch summit. 
I got a problem after building the TPLS, when running the cmake for Amanzi, from here: 
https://github.com/amanzi/amanzi/blob/ad5f430a8a97e6d2e18217e95d62b0fbe7cc7adb/bootstrap.sh#L1938

The option passed to the Amanzi CMAKE:
``` 
  -DMPI_EXEC_GLOBAL_ARGS:STRING=${mpi_exec_args} 
```

On summit the option: `mpi_exec_args=-a 1 -c 1 -g 1`
Which gives the cmake error: cmake "-g" option unknown. 
I tried to add " ", ' ' or \" \" around `mpi_exec_args` but nothing worked. I think we need to make a list? 

@david any idea? 

And beside that, to validate the MR, when I just remove this option it builds well and I can run on GPU.

File: CMakeLists.txt [hunk 2]
```diff
@@ -96,11 +96,13 @@ if ((ENABLE_Unstructured)  AND
                       "-D ENABLE_MESH_MSTK:BOOL=ON\n")
 endif()    
 
-if(ENABLE_KOKKOS_OPENMP)
-  set(CMAKE_CXX_FLAGS "-fopenmp")
+if(ENABLE_OpenMP)
+  find_package(OpenMP REQUIRED)
+  set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} ${OpenMP_C_FLAGS}")
```
[the remark this answers is not in the document]
Comment (JulienLoiseau) [reply CMakeLists.txt:101]: The CMAKE_C_FLAGS_DEBUG and CMAKE_C_FLAGS_RELEASE are added into CMAKE_C_FLAGS by cmake.

[the remark this answers is not in the document]
Comment (ecoon) [reply CMakeLists.txt:101]: No, CMAKE_C_FLAGS is the "master" set of flags, and CMAKE_C_FLAGS_DEBUG get appended to CMAKE_C_FLAGS if it is a Debug build (respectively Release).

File: CMakeLists.txt [hunk 3]
```diff
@@ -96,11 +96,13 @@ if ((ENABLE_Unstructured)  AND
...
+  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${OpenMP_CXX_FLAGS}")
 endif()
 
-if(ENABLE_KOKKOS_CUDA) 
+if(ENABLE_CUDA) 
     set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} \
```
[the remark this answers is not in the document]
Comment (JulienLoiseau) [reply CMakeLists.txt:106]: No, CMAKE_C_FLAGS is the "master" set of flags, and CMAKE_C_FLAGS_DEBUG get appended to CMAKE_C_FLAGS if it is a Debug build (respectively Release).

**Comment by lipnikov** [comment]
Bootstrap default options should be compatible with the code. Currently tpetra is off which seems inappropriate for this branch.

File: tools/cmake/AmanziTPL.cmake (hunk 1 above)
Comment (ecoon) [reply tools/cmake/AmanziTPL.cmake]: removed the comment

**Comment by ecoon** [comment]
> Bootstrap default options should be compatible with the code. Currently tpetra is off which seems inappropriate for this branch.

After much discussion, I've left epetra=True, tpetra=False as the defaults on the master branch.  To build TPLs for tpetra branch, one must at least --enable-tpetra, but that seems ok for the "expert developers" currently using that branch.  I've added an example to the bootstrap documentation of doing this with openmp.

**Comment by ecoon** [comment]
> I tried to add " ", ' ' or " " around `mpi_exec_args` but nothing worked. I think we need to make a list?

We've run into this in a bunch of ways now.  This is one way, but I've also tried to add a generic "cmake_extra_args" option that would let the user stuff more cmake configure flags into the bootstrap command line, and it doesn't work because of this same reason.

I've tried all sorts of stack overflow searching, bash script searching, and I just can't get it to work.  The one thing that does work is to change the line from:

    ${cmd_configure}

to 

    eval ${cmd_configure}

and then it works.  But this is considered a security flaw, so people don't like it.  Personally I think a bootstrap command is never injesting strings in a way that make me worry about it, but in general it can be problematic.  

I'm still waiting for @jd-moulton to figure a way around this, but I've given up.

**Comment by ecoon** [comment]
@JulienLoiseau  I'm trying to resolve the above merge conflicts, and am unsure of which is the correct patch to take.   Do you want to patch only Ifpack_Hypre now (what is on this PR branch) or do you need to patch both Ifpack_Hypre and Ifpack2_Hypre?

**Comment by JulienLoiseau** [comment]
@ecoon I have a separate file for ifpack and ifpack2 in this version, I was following a one patch per package. 
I just fixed the conflicts following that. 
If you prefer to have one for both ifpack and ifpack2 I can change that.

**Comment by ecoon** [comment]
No, that's great, I didn't see the extra patch.  If this is resolved, we'll wait for tests to pass here then merge this PR.

**Comment by jd-moulton** [comment]
I've tried several ways to work around the multiple argument problem in mpiexec, and I'm convinced the best (most robust and easiest way) to resolve this is to move to creating a cache file that is used by the cmake command.  We have used this approach successfully in some of the TPLs.  If we need this feature in this PR, I can work on it on the weekend.  It's technically my Friday off and we are headed to Albuquerque shortly. So I can't look at it today.

**Pull request merged**